libr/debug/p/native/xnu/xnu_debug.c

```diff
@@ -457,6 +478,215 @@ task_t pid_to_task (int pid) {
+
+		if (info.is_submap) nesting_depth++;
+		else address += size; n++;
+	}
+
+	return n;
+}
+
+static int get_bits () {
+#if __i386__ || __powerpc__
+	return R_SYS_BITS_32;
+#elif __x86_64__ || __mips__
+	return R_SYS_BITS_32 | R_SYS_BITS_64;
+#elif __arm__ || __aarch64__
+	return R_SYS_BITS_16 | R_SYS_BITS_32 | R_SYS_BITS_64;
+#else
+	return 0;
+#warning Unsupported architecture
+#endif
+}
+
+static int xnu_get_bits_with_sysctl (pid_t pid) {
+	int mib[4] = { CTL_KERN, KERN_PROC, KERN_PROC_PID, 0 };
+	size_t len = 4;
+	struct kinfo_proc kp;
+	size_t kinfo_size = sizeof (kp);
+
+	mib[3] = pid;
+	if (sysctl (mib, len, &kp, &kinfo_size, NULL, 0) == -1) {
+		perror ("sysctl");
+		return -1;
+  	} else {
+#if __arm__ || __aarch64__
+	return (kp.kp_proc.p_flag & P_LP64) ? 
+					R_SYS_BITS_16 | R_SYS_BITS_32 | R_SYS_BITS_64 
+					: R_SYS_BITS_16 | R_SYS_BITS_32;
+#else
+	return (kp.kp_proc.p_flag & P_LP64) ? 
+					R_SYS_BITS_32 | R_SYS_BITS_64 
+					: R_SYS_BITS_32;
+#endif
+  	}
+}
+
+/* RDebug doesnt set the bits. FIXME when that is done */
+#define SAME_BITNESS(proc_pid)\
+	 (get_bits () == xnu_get_bits_with_sysctl (proc_pid))
+
+static void get_mach_header_sizes(int *mach_header_sz, 
+									int *segment_command_sz) {
+#if __ppc64__ || __x86_64__
+	*mach_header_sz = sizeof(struct mach_header_64);
+	*segment_command_sz = sizeof(struct segment_command_64);
+#elif __i386__ || __ppc__
+	*mach_header_sz = sizeof(struct mach_header);
+	*segment_command_sz = sizeof(struct segment_command);
+#else
+#endif
+// XXX: What about arm?
+}
+
+#define COMMAND_SIZE(segment_count,segment_command_sz,\
+	thread_count,tstate_size)\
+	segment_count * segment_command_sz + thread_count * \
+	sizeof (struct thread_command) + tstate_size * thread_count
+
+static int xnu_build_header() {
+
+}
+
+static int xnu_dealloc_threads (RList *threads) {
+	RListIter *iter, *iter2;
+	xnu_thread_t *thread;
+	mach_msg_type_number_t thread_count;
+	thread_array_t thread_list;
+	kern_return_t kr = KERN_SUCCESS;
+
+	kr = task_threads (task_dbg, &thread_list, &thread_count);
+	if (kr != KERN_SUCCESS) {
+		perror ("task_threads");
+	} else {
+		r_list_foreach_safe (threads, iter, iter2, thread) {
+			mach_port_deallocate (mach_task_self (), thread->port);
+		}
+		vm_deallocate (mach_task_self (), (vm_address_t)thread_list,
+			thread_count * sizeof (thread_act_t));
+	}
+}
+
+static int xnu_prepare_corefile (RDebug *dbg, char *newcorefile) {
+
+	// XXX Due to platform specific implementation right now, only APPLE can get uid & gid.
+	char *corefile[MAXPATHLEN];
+	int corefile_fd;
+	RDebugInfo *info = r_debug_info (dbg, "");
+
+	/* If the newcorefile string is empty, setup a sane default based on system defaults */
+	if (strlen (newcorefile) == 0) {
+		snprintf (corefile, MAXPATHLEN, DEFAULT_COREFILE_DEST, dbg->pid);
+	} else {
+		strncpy (corefile, r_str_chop_ro (strdup (newcorefile)), MAXPATHLEN - 1);
+	}
+
+  	corefile_fd = open (corefile, O_RDWR | O_CREAT | O_EXCL, 0600);
```

Comment:
use api from r_util. i think that you should return an RBuffer instead of using raw open/write.. apis here because otherwise the sandbox (libr/util/sandbox.c) will be bypassed. just construct the data in RBuffer, and make RCore write the file in disk, memory or send it to the network or where the user wants.